src/core/grid_based_algorithms/OptionalCounter.hpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2020 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdint>
+
+#include <utils/Counter.hpp>
+
+/** Re-implementation of a boost::optional for a RNG counter.
+ *
+ *  Workaround for a compiler error with Clang 9.0, boost 1.71
+ *  and CUDA 10.1 (see espressomd/espresso#3650).
+ */
+class OptionalCounter {
+private:
+  Utils::Counter<uint64_t> m_counter;
+  bool m_initialized;
+
+public:
+  OptionalCounter() : m_counter{}, m_initialized(false) {}
+  OptionalCounter(Utils::Counter<uint64_t> &counter)
+      : m_counter(counter), m_initialized(true) {}
+  OptionalCounter &operator=(Utils::Counter<uint64_t> &counter) {
```

Comment:
`const &` or take by value, in which case you can drop the self-assignment check